### PR TITLE
[SDK-222] [no connection serial ] + implement missing internal presence map + presencequeue

### LINKED ISSUE #255: RTP2, RTP17, RTP19: PresenceMap
Write spec tests for [RTP2](https://docs.ably.io/client-lib-development-guide/features/#RTP2), [RTP17](https://docs.ably.io/client-lib-development-guide/features/#RTP17), [RTP19](https://docs.ably.io/client-lib-development-guide/features/#RTP19), ensure they pass and that the affected API conforms to [the 1.2 API design](https://github.com/ably/ably-go/issues/197).

┆Issue is synchronized with this [Jira Story](https://ably.atlassian.net/browse/SDK-222) by [Unito](https://www.unito.io)

### PR DESCRIPTION
**Pull request by sacOO7**
- Fixed #255

### CODE REVIEW

**Comment by sacOO7** [comment]
TODO - Check for channel attach while publishing presence msg

File: ably/export_test.go
Comment (lmars) [inline ably/export_test.go:230]: If this map might get mutated after this call returns, then holding the lock just to return the map isn't enough, the map itself needs to be thread safe.

File: ably/proto_protocol_message_test.go
Comment (lmars) [inline ably/proto_protocol_message_test.go:27]: The comment is now incorrect, this is a two-element map.

File: ably/realtime_channel.go
Comment (lmars) [inline ably/realtime_channel.go:550]: This comment doesn't match the method.

File: ably/realtime_channel.go
Comment (lmars) [inline ably/realtime_channel.go:558]: Given this function never returns an error, perhaps drop the `err` return value?

File: ably/realtime_presence.go
Comment (lmars) [inline ably/realtime_presence.go:150]: Is there any context as to why this timeout is set to 2 minutes?

File: ably/realtime_presence.go
Comment (lmars) [inline ably/realtime_presence.go:172]: This looks susceptible to deadlock, where does this get locked? Locks and unlocks should really always appear in pairs close to each other.

Comment (sacOO7) [reply ably/realtime_presence.go:172]: This is also done because of `syncWait` method, you can check it's usage in comments
https://github.com/ably/ably-go/blob/e66ad6a1bfa0e0e48dccb7c1afd0c282b568d1c2/ably/realtime_presence.go#L93-L99

File: ably/realtime_presence.go
Comment (sacOO7) [reply ably/realtime_presence.go:150]: Honestly there's no context as such, I just didn't want goroutine to get blocked!
Your suggestions will be helpful here !!

File: ably/export_test.go
Comment (sacOO7) [reply ably/export_test.go:215]: Yeah, `dismiss` returns queuedMessages and empties internal queue

File: ably/export_test.go
Comment (sacOO7) [reply ably/export_test.go:230]: actually these methods are only used for `testing`, are not used anywhere else.
All methods defined under `export_test.go` are available only for tests.

File: ably/realtime_presence.go
Comment (SimonWoolf) [inline ably/realtime_presence.go:162]: The naming of `isAttachWithoutMessageLoss` seems misleading. That'll be set to true in any situation where you need to re-enter your members, which includes both situations where you do and don't have message loss

File: ably/realtime_channel.go
Comment (SimonWoolf) [inline ably/realtime_channel.go:817]: This conditional looks kinda weird. You'll be in this `else` clause if either of the two conditions in the if fail, but the stuff in the else clause only makes sense if it was the first condition that failed -- eg what if you get an attached with resumed=true while you were already attached? seems like it should be
```
		if c.state == ChannelStateAttached {
			if !msg.Flags.Has(flagResumed) {
				c.emitErrorUpdate(newErrorFromProto(msg.Error), false) // RTL12
				c.Presence.onAttach(msg, false)
			}
		} else {
			c.Presence.onAttach(msg, true)
			c.setState(ChannelStateAttached, newErrorFromProto(msg.Error), msg.Flags.Has(flagResumed))
		}
```

File: ably/realtime_conn.go
Comment (SimonWoolf) [inline ably/realtime_conn.go:606]: why does `closed` need to be special-cased? can't it be handled by `connStateError(state)` like all the others?

File: ably/realtime_conn.go
Comment (SimonWoolf) [inline ably/realtime_conn.go:619]: :+1: good job spotting that bug

File: ably/realtime_presence.go
Comment (SimonWoolf) [inline ably/realtime_presence.go:269]: as discussed above -- I'd get rid of this

File: ably/realtime_presence.go
Comment (sacOO7) [reply ably/realtime_presence.go:162]: Let's say, if you `attached` channel when previous state was `detached` or `attaching`, do we get `resume` as false in such a case?

Comment (sacOO7) [reply ably/realtime_presence.go:162]: So, we should call it `isFreshAttach` wdyt?

File: ably/realtime_presence.go
Comment (sacOO7) [reply ably/realtime_presence.go:269]: Not sure why we even added this in the first place. I am sure there was a concrete reason

Comment (sacOO7) [reply ably/realtime_presence.go:269]: You can see in `ably-dotnet`, we have a `realtimechannel` method called `OnError`.
https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/RealtimeChannel.cs#L559-L567
This gets called whenever there is any error registered on a channel 
https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/RealtimeChannel.cs#L628

You can see there is an error delegate registration here 
https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/RealtimeChannel.cs#L57C44-L57C44

Comment (sacOO7) [reply ably/realtime_presence.go:269]: `OnError` method usages
1. Presence -> https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/Presence.cs#L610
2. Sync message -> https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/ChannelMessageProcessor.cs#L148
3. Channel state errors -> https://github.com/ably/ably-dotnet/blob/c2a514ad2cc9401547c8c6ccb7c23b87101131a5/src/IO.Ably.Shared/Realtime/RealtimeChannel.cs#L628

File: ably/realtime_channel.go
Comment (sacOO7) [reply ably/realtime_channel.go:553]: I believe we do have `OnError` public method for `ably-js` and `ably-dotnet`.

File: ably/realtime_presence.go
Comment (sacOO7) [reply ably/realtime_presence.go:269]: Let me know if we should keep it. If this is not part of the spec, then there's no point in keeping the public method.

Comment (sacOO7) [reply ably/realtime_presence.go:269]: Seems we need to do something like this 
https://github.com/ably/ably-java/blob/72992c943498ee2ebefb8acec303718ed7f35ae4/lib/src/main/java/io/ably/lib/realtime/ChannelBase.java#L1336C1-L1339C6

Comment (sacOO7) [reply ably/realtime_presence.go]: Okay, I have removed this particular piece of code. 
Let me know if we planning to add `OnError` method to the spec. I think this will be a very specific way to listen to channel-specific errors as we do the same in `ably-dotnet`.

File: ably/realtime_conn.go
Comment (sacOO7) [reply ably/realtime_conn.go:606]: Actually 
https://github.com/ably/ably-go/blob/e66ad6a1bfa0e0e48dccb7c1afd0c282b568d1c2/ably/state.go#L55-L60 map doesn't contain `errClosed`. 
At the same time, we can't update the map to include `errClosed` since it can be a graceful close from the client side. 
If we update the map with `errClosed`, other tests start failing where we don't expect error when client goes into closed state, like when we call `client.close()` and `conn` state changes to `closed` with error.
This error is specific to our usecase, so I didn't include it into the map

File: ably/proto_protocol_message_test.go
Comment (sacOO7) [reply ably/proto_protocol_message_test.go:27]: updated

File: ably/realtime_conn.go
Comment (sacOO7) [reply ably/realtime_conn.go:619]: thanks

File: ably/export_test.go
Comment (lmars) [reply ably/export_test.go:215]: Ok, but isn't that _not_ resending them, instead emptying the queue and ack'ing them?

File: ably/export_test.go
Comment (lmars) [reply ably/export_test.go:230]: Ok, but even in the tests we should be concerned about making things thread safe, otherwise we'll get failing tests due to races being detected.

File: ably/realtime_presence.go
Comment (lmars) [reply ably/realtime_presence.go:150]: Perhaps it should keep retrying whilst the channel is in a state where we can expect the operation to eventually succeed (i.e. it's not in a terminal state)?

File: ably/realtime_presence.go
Comment (lmars) [reply ably/realtime_presence.go:172]: Ok, that's really not a good use of a mutex, a mutex should be used to synchronise access to shared memory, not to synchronise clients waiting for the result of a network call (i.e. here waiting for a presence sync to complete). For example, what happens if the sync doesn't complete because the channel became detached? Does the mutex stay locked, and all calls to `syncWait` block until the program exits?

It's typical to use the closing of an empty struct channel to unblock multiple callers when some event has happened, and to select on it in a cancelable way:

```go
// a field on RealtimePresence
syncDone := make(chan struct{})

// when the sync has completed
close(p.syncDone)

// wait for the channel to close in syncWait
func (p *RealtimePresence) syncWait(ctx context.Context) error {
select {
case <-p.syncDone:
	return nil
case <-ctx.Done():
	return ctx.Err()
}
```

File: ably/realtime_presence.go
Comment (sacOO7) [reply ably/realtime_presence.go:150]: I just don't want goroutine to go into a dangling state : (
Sure, it can work.

File: ably/export_test.go
Comment (sacOO7) [reply ably/export_test.go:215]: Yeah, I will update message for the same

File: ably/realtime_presence.go
Comment (sacOO7) [reply ably/realtime_presence.go:172]: sure, we can update

File: ably/realtime_channel.go
Comment (SimonWoolf) [reply ably/realtime_channel.go:806]: > I think RTL12 should mention explicitly, if attached is received on already attached channel and resume set to true, then it should do nothing

It already does say that. "If and only if the resumed flag is false, this should result in the channel emitting an UPDATE event with a ChannelStateChange object". "If _and only if_"

File: ably/realtime_channel.go
Comment (sacOO7) [reply ably/realtime_channel.go]: Sorry, i should have updated it. Only dotnet has it

File: ably/realtime_conn.go
Comment (SimonWoolf) [reply ably/realtime_conn.go:606]: > At the same time, we can't update the map to include errClosed since it can be a graceful close from the client side.

I don't follow. If the connection is closed, then most operations you try to do to it (other than reconnect) should fail with a 'connection closed' error, similarly to if it's failed. if connclosed is missing from that then either every place you call is going to need to check if the conn is closed and handle that explicitly, or you have a bug, no?

> E.g. call to client.close(), and conn state changes to closed with error.

No it doesn't because connection state changes don't have errors. They have `reason`s. The reason might be 'the connection was closed', that's not an error, it's telling you the connection is in that state

File: ably/realtime_presence.go
Comment (SimonWoolf) [reply ably/realtime_presence.go]: > You can see in ably-dotnet, we have a realtimechannel method called OnError.

I am not a dot.net developer. But my understanding is that if it's not part of the realtime channel interface then it's not part of the public api. It's an internal method, used by the library itself. It's annotated 'public' because it can be called by other parts of the library, but that doesn't make it part of the actual public API.

> Let me know if we planning to add OnError method to the spec.

We're not.

> [Seems we need to do something like this](https://github.com/ably/ably-java/blob/72992c943498ee2ebefb8acec303718ed7f35ae4/lib/src/main/java/io/ably/lib/realtime/ChannelBase.java#L1336C1-L1339C6)

[We already have that](https://github.com/ably/ably-go/blob/fix/presencemap/ably/realtime_channel.go#L995)

Comment (sacOO7) [reply ably/realtime_presence.go]: Got it 👍

File: ably/realtime_presence.go
Comment (SimonWoolf) [reply ably/realtime_presence.go:150]: > To put another way, what if the enterClient call is failing due to an unreliable network for 3 minutes, but then it's ok again, shouldn't this then succeed rather than have give up after 2 minutes?

If a message is sent to the connection while it's in a `connected` state, then eventually either it will successfully send or the connection will detect a network issue and become disconnected. If the latter happens, then when it reconnects, the channel will re-attach, which will trigger `onAttach()` and `presence.enterMembers()` to happen again at that time. So we shouldn't need an old enterMembers from some previous attach cycle living on.

File: ably/realtime_conn.go
Comment (sacOO7) [reply ably/realtime_conn.go:606]: I believe it's called `errorReason` =>
https://github.com/ably/ably-go/blob/e66ad6a1bfa0e0e48dccb7c1afd0c282b568d1c2/ably/realtime_conn.go#L48-L51
and same is copied into reason
https://github.com/ably/ably-go/blob/e66ad6a1bfa0e0e48dccb7c1afd0c282b568d1c2/ably/realtime_conn.go#L1001-L1006

There are tests which doesn't treat connection closed as connection failure as such. I mean ideally they shouldn't.
Because, `client.close()` initiates it and gracefully closes the connection accordingly.

File: ably/realtime_presence.go
Comment (lmars) [reply ably/realtime_presence.go:150]: Sorry yes, I'm proposing `context.Background()`, since if I understand Simon correctly, the call will either block until it succeeds, or return an error if the client becomes disconnected, so in effect there is a timeout on this operation based on how long it takes the client to detect a disconnection, there's no need to specify an arbitrary timeout here too.

File: ably/realtime_conn.go
Comment (sacOO7) [reply ably/realtime_conn.go:606]: Do we have a spec item that specifies to set a reason when client explicitly sends a `close` message and  connection is closed by server

File: ably/realtime_presence.go
Comment (sacOO7) [reply ably/realtime_presence.go:150]: Same will happen with `cancelledContext` too. If it fails to send message due to `disconnection`, for next channel attach it will automatically send the message. We don't really care about message failure in current case. Because, it will only fail when channel is in some other state ( maybe because of disconnection ), for next attach, we can enter members again.
There's no need to wait really

File: ably/realtime_presence.go
Comment (sacOO7) [reply ably/realtime_presence.go:179]: Updated as a part of -> https://github.com/ably/ably-go/pull/621/commits/2740853cd88a58e6b0b6332904e61efcb3c4912c

File: ably/export_test.go
Comment (sacOO7) [reply ably/export_test.go]: updated -> https://github.com/ably/ably-go/pull/621/commits/31b739a380c961ff176671093579c8928e4b9792

File: ably/export_test.go
Comment (sacOO7) [reply ably/export_test.go]: updated -> https://github.com/ably/ably-go/pull/621/commits/31b739a380c961ff176671093579c8928e4b9792

File: ably/realtime_presence.go
Comment (lmars) [reply ably/realtime_presence.go]: Ok thanks, can we get rid of `syncMtx` altogether?

Comment (sacOO7) [reply ably/realtime_presence.go]: we did

File: ably/realtime_channel.go
Comment (sacOO7) [reply ably/realtime_channel.go:808]: ```The library must never emit a ChannelState ChannelEvent for a state equal to the previous state``` means it should emit `UpdateEvent` for prevState = newState` instead of `ChannelEvent`. That's what I understood by it.

Comment (sacOO7) [reply ably/realtime_channel.go:808]: As per RTL2f =>
```
When a channel ATTACHED ProtocolMessage is received, the ProtocolMessage may contain a RESUMED bit flag indicating that the channel has been resumed. The corresponding ChannelStateChange (either ATTACHED per RTL2a, or UPDATE per RTL12) will contain a resumed boolean attribute with value true if the bit flag RESUMED was included ...
```
I feel there's a gray area here. It doesn't mention if it's a ATTACHED on non-attached channel or ATTACHED on already attached channel. I can only assume if it's ATTACHED on already ATTACHED with resume flag as true, as per RTL2g, it should emit update event on the channel.
RTL12 only describes ATTACHED on already ATTACHED with resume flag as false, other specs are implicitly pointing towards setting a state and emitting update event for the same.

Comment (sacOO7) [reply ably/realtime_channel.go:808]: I have updated the code according to your suggestion.
But, I would be honest, I had to read related spec multiple times again and again for this. Still, as you can see, understanding wasn't reflected in the code.
Maybe I will update the spec here https://github.com/ably/specification/pull/167 and we can resume discussion for the same.

File: ably/realtime_presence.go
Comment (sacOO7) [reply ably/realtime_presence.go:151]: Update -> `context.Background()` does seem to be a sensible approach, but is there a way this can go into a lock type of a situation where client is connected for a long time and ack is not received for the entered presence.

Comment (sacOO7) [reply ably/realtime_presence.go:151]: That's why I tried putting a `timeout` here, where all `innerPresenceMembers` will be processed without one waiting for another

Comment (lmars) [reply ably/realtime_presence.go:151]: Ok, I leave it to your judgement, I'd probably just use `context.Background()`

File: ably/realtime_presence.go
Comment (lmars) [reply ably/realtime_presence.go]: Yes sorry, I think I was looking at an outdated diff 👍

**Review by lmars — APPROVED** [review]
This looks ok from my perspective, but I defer to @SimonWoolf to review the changes he requested

File: ably/realtime_channel.go
Comment (sacOO7) [reply ably/realtime_channel.go:808]: I have added explicit commit to the specification PR https://github.com/ably/specification/pull/167/commits/248bbc63a29e7393e56c2b72b92ab5097b628e30

Comment (sacOO7) [reply ably/realtime_channel.go:808]: Since the change has been made as per suggestion, I am closing the issue

**Comment by sacOO7** [comment]
@lmars merging this PR into integration-2.0, since one more review required for recovery spec related commits. I will re-raise @SimonWoolf comments there

**Pull request merged**
